### PR TITLE
Ranger

### PR DESCRIPTION
**Pull request by alex-novikov-1990**

### CODE REVIEW

File: ranger/src/bmt-db.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                            FOR t IN a_b.transactions
+                                RETURN t.transaction_id
+                    ),
+                }
+        `);
+
+        return await cursor.all() as Block[];
+    }
+
+    async find_existant_transaction_ids(ids: string[]): Promise<string[]> {
```
Comment (melsomino) [inline ranger/src/bmt-db.ts:29]: existing-же

File: ranger/src/chain-ranges-verification-db.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            throw new Error("Chain ranges summary not found");
+        }
+        return result;
+    }
+
+    async update_verifyed_boundary(veryfied_master_seq_no: number): Promise<void> {
+        const chain_order_boundary = toU64String(veryfied_master_seq_no + 1);
+
+        await this.database.query(aql`
+            UPDATE "summary"
```
Comment (melsomino) [inline ranger/src/chain-ranges-verification-db.ts:31]: Maybe use UPSERT here?

File: ranger/src/ranger.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private ensure_chain_range_on_blocks(blocks: Block[]): void {
+        const blocks_without_co = blocks.filter(b => !b.chain_order);
+        if (blocks_without_co.length > 0) {
+            throw new Error(`There is no chain range on blocks: ${blocks_without_co.map(b => b.id).join(", ")}`);
+        }
+    }
+
+    private async find_existant_transaction_ids(ids: string[]): Promise<string[]> {
```
Comment (melsomino) [inline ranger/src/ranger.ts:69]: existing

File: ranger/src/ranger.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            const delayPromise = delay(this.database_polling_period);
+            if (block_ids_to_check.length > 0) {
+                const blocks = await this.bmt_db.find_blocks_by_ids(block_ids_to_check);
+                this.ensure_chain_range_on_blocks(blocks);
+                transcation_id_to_check = transcation_id_to_check.concat(...blocks.map(b => b.transaction_ids));
+                block_ids_to_check = block_ids_to_check.filter(id => !blocks.find(b => b.id == id));
+            }
+            if (transcation_id_to_check.length > 0) {
+                const transaction_ids = await this.find_existant_transaction_ids(transcation_id_to_check);
+                transcation_id_to_check = transcation_id_to_check.filter(id => !transaction_ids.find(id2 => id2 == id));
```
Comment (melsomino) [inline ranger/src/ranger.ts:51]: const transaction_ids = new Set(await this.find_existant_transaction_ids(transcation_id_to_check));
transcation_id_to_check = transcation_id_to_check.filter(id => !transaction_ids.has(id));

File: ranger/src/ranger.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        let block_ids_to_check = [chain_range.master_block.id, ...chain_range.shard_block_ids];
+        let transcation_id_to_check = [] as string[];
+
+        while (true) {
+            const delayPromise = delay(this.database_polling_period);
+            if (block_ids_to_check.length > 0) {
+                const blocks = await this.bmt_db.find_blocks_by_ids(block_ids_to_check);
+                this.ensure_chain_range_on_blocks(blocks);
+                transcation_id_to_check = transcation_id_to_check.concat(...blocks.map(b => b.transaction_ids));
+                block_ids_to_check = block_ids_to_check.filter(id => !blocks.find(b => b.id == id));
```
Comment (melsomino) [inline ranger/src/ranger.ts:47]: const block_ids = new Set(blocks.map(x => x.id));
block_ids_to_check = block_ids_to_check.filter(id => !block_ids.has(id));

**Review by melsomino — CHANGES_REQUESTED** [review]
see changes

File: ranger/src/chain-ranges-verification-db.ts (hunk 2 above)
Comment (alex-novikov-1990) [reply ranger/src/chain-ranges-verification-db.ts:31]: The ranger requires existing summary to start. So insert is impossible by design.

File: ranger/src/bmt-db.ts (hunk 1 above)
Comment (alex-novikov-1990) [reply ranger/src/bmt-db.ts:29]: Fixed

File: ranger/src/ranger.ts (hunk 3 above)
Comment (alex-novikov-1990) [reply ranger/src/ranger.ts:69]: Fixed

File: ranger/src/ranger.ts (hunk 4 above)
Comment (alex-novikov-1990) [reply ranger/src/ranger.ts:51]: Done

File: ranger/src/ranger.ts (hunk 5 above)
Comment (alex-novikov-1990) [reply ranger/src/ranger.ts:47]: There are less than 1000 blocks per range (averaging to 100 at high load of blockchain). So the impact on performance will be insignificant, but the code of lines between will be cluttered (`blocks.map` and `blocks.filter` parts of code).

**Review by melsomino — APPROVED** [review]

**Pull request merged**